### ISSUE TITLE
New rule: `use-tag` to enforce which outside variables can go into closures (depending on @use)

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#11061]
**Please describe what the rule should do:**

I'd like to see something like the following:

```js
var outer1 = 1;
var outer2 = 2;
var outer3 = 3;

/**
* @use outer1|outer2
*/
function myClosure () {
    outer1; // ok
    outer2; // ok
    outer3; // Flagged as problematic
}
```

By default, it may include built-ins or `this`.

There'd be a `use-tag` rule:

1. Indicate `requireUse` (default: `{all: {required: "off", enforce: {when: "explicit", defaultTypes: ["this",  "builtins"]}}`); would either be an object with the single key "all" or an object with any or all of these keys: `MethodDefinition`, `FunctionDeclaration`, `nonGlobals`, `ArrowFunction`, `FunctionExpression`.

The top level keys (either of which could be omitted) would be:

1. `required` (default: "off")
    1. `always` - Require the `@use` tag (explicit or implicit) for all functions of the given type.
    2. `nonGlobals` - Require the `@use` tag for functions of the given type defined outside of the global scope
    3. `globals` - Require the `@use` tag for functions of the given type defined at the global scope (usable for indicating the built-ins employed).
    4. `never` - Prohibit its use
    5. `off` - No requirement to have `@use` for functions (other rules below may be used, however)
2. `enforce`
    1. `when` (default: "explicit")
        1. `explicit` - Would prevent non-local variables not explicitly added within the `@use` whitelist
        1. `implicit` - Besides preventing non-local variables when `@use` is explicitly added, this option will also treat all functions by default as though a null `@use` were applied (i.e., no outside variables can be used within such functions, unless they explicitly add a `@use`).
        1. While `explicit` is the default, for debugging it might perhaps be helpful to have an `off` setting which would allow requiring the `@use` tag but not actually enforcing its rules even for explicit cases.
    2. `defaultTypes` (default: `["this",  "builtins"]`) - An array of "this" and/or "builtins". Such array items indicate the enforcement is implied with no need for these to be explicitly added to `@use`. Perhaps when "implicit" was used, allowing for either "implicit" or "explicit", there could be a separate `explicitDefaultTypes` to allow for configuration pertaining only to "explicit" `@use` tags.
3. `valid` - Flag inclusion of `@use` without variables, while permitting a keyword like `null` to signal that no items should be permitted. Ensure the `@use` tag follows the pipe-separated JavaScript identifiers syntax (and besides `null`, no keywords besides `this` may be used) and prohibits use of parameter names. Where `require` or `enforce` are used, this is automatically applied.
4. `overrideNoShadow` (default: true) - When `@use` is in place, this option should allow shadowing for any variable names within the function except for variables whitelisted within `@use`.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[X] Other (please specify:)

The rule would ensure that no outside variables were accidentally used (when a local declaration should have been made instead).

Moreover, this rule with its `overrideNoShadow` option also allows the convenience of allowing shadowing for functions using `@use` implicitly or explicitly (for variables not specified within `@use`). The concern for a variable from the outer scope being mistakenly used can be minimized here  because the inclusion of `@use` would suggest that non-whitelisted items were not being used (and this could be checked).

This rule could also facilitate better modularization of functions, avoiding too many contextual dependencies (and preventing the closure being moved externally if necessary).

**Provide 2-3 code examples that this rule will warn about:**

Implicit null `@use` enforcement:

```js
/* eslint use-tag: ["error", {all: {required: "off", enforce: {when: "implicit" defaultTypes: ["this",  "builtins"]}}]  */
var outer1 = 1;

function myClosure () {
    outer1; // Flagged as problematic
}
```

```js
/* eslint use-tag: ["error", {all: {required: "off", enforce: {when: "implicit" defaultTypes: []}}]  */

function myClosure () {
    this; // Flagged as problematic as defaultTypes does not whitelist `this`
}

function myClosure2 () {
    Array; // Flagged as problematic as defaultTypes does not whitelist `builtins`
}
```

Missing `@use`:
```js
/* eslint use-tag: ["error", {all: {required: "always"}]  */

function myClosure () {
}
```

With `overrideNoShadow` disabled, the `no-shadow` rules applies as usual.

```js
/* eslint no-shadow: ["error"] */
/* eslint use-tag: ["error", {all: {overrideNoShadow: false, required: "off", enforce: {when: "implicit" defaultTypes: []}}]  */
var outer1 = 1;

/**
* @use outer2
*/
function myClosure () {
    // This would normally be permitted because `@use` hasn't indicated
    //   expecting `outer1`, but with `overrideNoShadow` disabled, it is
    //   disallowed
    const outer1 = 5;
}

```

**Why should this rule be included in ESLint (instead of a plugin)?**

Because control of what goes into closures ought to be a basic functionality.

While the tag `@use` is not standard JSDoc, it would fill an enormous need (and JSDoc allows extensions).

**Are you willing to submit a pull request to implement this rule?**

Not at this time.

**Extra info**

`@use` also leverages the terminology of PHP in its ["use" language construct](http://www.php.net/manual/en/functions.anonymous.php#example-164).

**Comment by nzakas** [comment]
Thanks for the proposal. Unfortunately, it looks like the team doesn't feel that this would make a good addition to the ESLint core. As mentioned previously, you can still create your own custom rule to fulfill this purpose if you find it useful for your code.
